### PR TITLE
Allow over-weight collective proposals to be closed

### LINKED ISSUE #6160: Allow over-weight collective proposals to be closed
Collective may incorrectly introduce proposals that are too big to call `close` on.

Such proposals can never be accepted or dispatched by the pallet per the limits of our weight system. Instead, these kinds of proposals will need to be routed through democracy to then be pushed to the scheduler.

However, we still need a way to close such a proposal.

### PR DESCRIPTION
**Pull request by shawntabrizi**
Closes #6160 

## Problem Statement

Given the new rules enacted by our weight system, there are a set of dispatchable calls which are simply not allowed by our system due to weight restrictions. One simple example is a "full block call" like `set_code`, but a more unique example may be some "fibonacci" extrinsic whose weight grows based on the input of the function. In our current weight system, such a function would in some cases be un-callable because it would alone trigger too much computation.

There is currently no mechanism to allow such a call to execute if the expected origin is something other than `Root`. Only `Root` calls are allowed to go past our weight limits through the use of the Scheduler.

## Effect on Collective

The Collective pallet allows users to submit proposals. When approved, these proposals will be dispatched and executed. However, it is completely allowable that a user submits a proposal that cannot be dispatched due to weight restrictions.

If such a proposal is made, the `close` transaction would be rejected by the node since it's weight is always larger than the proposal being dispatched. This error was appearing no matter if the proposal would be approved or disapproved, forcing such proposals to be stuck in the system storage.

## Solution

This PR splits the weight calculation when a proposal is accepted versus when it is rejected while calling `close`.

When a proposal is rejected, we do not care about the weight or length information passed by the caller since the proposal will never be read from storage or dispatched. Instead, we can simply bypass any checks on the weight/len input parameters and just clean up the storage. This means we can call `close` with weight = 0 and len = 0 in the case of rejection. cc @jacogr 

When a proposal is accepted, we do care about this weight and length information, and we keep doing the same checks that are already present. It is possible that users will still propose some call that would not be possible through the collective pallet. Such calls will need to be rejected by the collective, and closed using the logic described above.

If for some crazy reason collective votes to accept a proposal that cannot be accepted, and the voting period passes, we introduce a root call `disapprove_proposal` that allows us to clean up the proposal information. This of course is not the best thing, but we are already dealing with a collective doing weird things.

### CODE REVIEW

File: frame/collective/src/lib.rs [hunk 1]
```diff
@@ -684,19 +690,46 @@ decl_module! {
...
 				proposal_weight_bound
 			)?;
-			Self::deposit_event(RawEvent::Closed(proposal, yes_votes, no_votes));
-			let finalize_weight = Self::finalize_proposal(approved, seats, voting, proposal, p);
+			Self::deposit_event(RawEvent::Closed(proposal_hash, yes_votes, no_votes));
+			let finalize_weight = if approved {
+				Self::do_approve_proposal(seats, voting, proposal_hash, proposal)
+			} else {
+				Self::do_disapprove_proposal(proposal_hash)
+			};
```
Comment (apopiak) [inline frame/collective/src/lib.rs:703]: The semantics here are different from the ones above. AFAICT this code has the old behavior (of getting stuck for overweight proposals) if the proposal exits the voting period.
You'll want to reorder like so (pseudo-code):
```
let w = if approved {
    let proposal = validate_and_get(hash)?;
    do_approve(proposal)
} else {
    do_disapprove(hash)
};
deposit_event(Closed);
```
I would suggest adding a test for the new behavior after end of the voting period as well.

Comment (shawntabrizi) [reply frame/collective/src/lib.rs]: Take a second look please!

File: frame/collective/src/lib.rs [hunk 2]
```diff
@@ -1114,36 +1124,56 @@ mod tests {
 	}
 
 	#[test]
-	fn proposal_weight_limit_works() {
+	fn proposal_weight_limit_works_on_approve() {
 		new_test_ext().execute_with(|| {
 			let proposal = Call::Collective(crate::Call::set_members(vec![1, 2, 3], None, MAX_MEMBERS));
```
Comment (apopiak) [inline frame/collective/src/lib.rs:1129]: why not set the prime here?

Comment (shawntabrizi) [reply frame/collective/src/lib.rs:1129]: This is the proposal, not an actual call.

This was confusing to me in the test as well. I need the prime to ACTUALLY be set when we do the vote, whereas this is the proposal being voted on.

Comment (apopiak) [reply frame/collective/src/lib.rs:1129]: Yeah, now that you mention it, I chose that call, maybeshould've picked something else :grimacing:

**Pull request merged**